# Part 1: Classifying movie reviews

IMDB dataset

In [16]:
import numpy as np
from keras import models, layers, optimizers, losses, metrics

In [1]:
from keras.datasets import imdb

(train_data, train_labels), (test_data, test_labels) = imdb.load_data(num_words=10000)

Using TensorFlow backend.


In [4]:
train_data[0][:10]

[1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65]

#### Preparing the data

In [8]:
def vectorize_sequences(sequences, dimension=10000):
    results = np.zeros((len(sequences), dimension))
    for i, sequence in enumerate(sequences):
        results[i, sequence] = 1.
    return results

x_train = vectorize_sequences(train_data)
x_test = vectorize_sequences(test_data)

In [9]:
y_train = np.asarray(train_labels).astype('float32')
y_test = np.asarray(test_labels).astype('float32')

In [10]:
y_train[:10]

array([1., 0., 0., 1., 0., 0., 1., 0., 1., 0.], dtype=float32)

In [13]:
x_train[1]

array([0., 1., 1., ..., 0., 0., 0.])

### Building the network

#### 3 hidden layers

In [17]:
model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss=losses.binary_crossentropy,
              metrics=[metrics.binary_accuracy])

In [24]:
x_val = x_train[:5000]
partial_x_train = x_train[5000:]
y_val = y_train[:5000]
partial_y_train = y_train[5000:]

In [25]:
history = model.fit(x_train, y_train, epochs=10, batch_size=512, validation_data=(x_val, y_val))
results = model.evaluate(x_test, y_test)

Train on 25000 samples, validate on 5000 samples
Epoch 1/10
25000/25000 [==============================] - 18s 726us/step - loss: 0.1928 - binary_accuracy: 0.9289 - val_loss: 0.1860 - val_binary_accuracy: 0.9294
Epoch 2/10
25000/25000 [==============================] - 16s 639us/step - loss: 0.1623 - binary_accuracy: 0.9421 - val_loss: 0.1181 - val_binary_accuracy: 0.9632
Epoch 3/10
25000/25000 [==============================] - 12s 498us/step - loss: 0.1393 - binary_accuracy: 0.9489 - val_loss: 0.0991 - val_binary_accuracy: 0.9718
Epoch 4/10
25000/25000 [==============================] - 15s 595us/step - loss: 0.1232 - binary_accuracy: 0.9561 - val_loss: 0.0862 - val_binary_accuracy: 0.9762
Epoch 5/10
25000/25000 [==============================] - 12s 498us/step - loss: 0.1073 - binary_accuracy: 0.9622 - val_loss: 0.0756 - val_binary_accuracy: 0.9796
Epoch 6/10
25000/25000 [==============================] - 11s 454us/step - loss: 0.1007 - binary_accuracy: 0.9637 - val_loss: 0.0645 - v

In [37]:
print('Test accuracy: {}'.format(results[1]))

Test accuracy: 0.85732


#### 2 hidden layers

In [30]:
model_2 = models.Sequential()
model_2.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model_2.add(layers.Dense(16, activation='relu'))
model_2.add(layers.Dense(1, activation='sigmoid'))

model_2.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss=losses.binary_crossentropy,
              metrics=[metrics.binary_accuracy])

In [31]:
history_2 = model_2.fit(x_train, y_train, epochs=10, batch_size=512, validation_data=(x_val, y_val))
results_2 = model_2.evaluate(x_test, y_test)

Train on 25000 samples, validate on 5000 samples
Epoch 1/10
25000/25000 [==============================] - 26s 1ms/step - loss: 0.4825 - binary_accuracy: 0.8094 - val_loss: 0.3077 - val_binary_accuracy: 0.9090
Epoch 2/10
25000/25000 [==============================] - 18s 715us/step - loss: 0.2716 - binary_accuracy: 0.9068 - val_loss: 0.2088 - val_binary_accuracy: 0.9340
Epoch 3/10
25000/25000 [==============================] - 17s 696us/step - loss: 0.2079 - binary_accuracy: 0.9253 - val_loss: 0.1649 - val_binary_accuracy: 0.9478
Epoch 4/10
25000/25000 [==============================] - 16s 629us/step - loss: 0.1741 - binary_accuracy: 0.9390 - val_loss: 0.1426 - val_binary_accuracy: 0.9550
Epoch 5/10
25000/25000 [==============================] - 16s 636us/step - loss: 0.1509 - binary_accuracy: 0.9475 - val_loss: 0.1269 - val_binary_accuracy: 0.9576
Epoch 6/10
25000/25000 [==============================] - 17s 695us/step - loss: 0.1331 - binary_accuracy: 0.9532 - val_loss: 0.1074 - val

In [38]:
print('Test accuracy: {}'.format(results_2[1]))

Test accuracy: 0.86496


#### 1 hidden layer

In [33]:
model_1 = models.Sequential()
model_1.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model_1.add(layers.Dense(1, activation='sigmoid'))

model_1.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss=losses.binary_crossentropy,
              metrics=[metrics.binary_accuracy])

In [34]:
history_1 = model_1.fit(x_train, y_train, epochs=10, batch_size=512, validation_data=(x_val, y_val))
results_1 = model_1.evaluate(x_test, y_test)

Train on 25000 samples, validate on 5000 samples
Epoch 1/10
25000/25000 [==============================] - 19s 774us/step - loss: 0.4323 - binary_accuracy: 0.8325 - val_loss: 0.2869 - val_binary_accuracy: 0.9136
Epoch 2/10
25000/25000 [==============================] - 14s 575us/step - loss: 0.2634 - binary_accuracy: 0.9103 - val_loss: 0.2130 - val_binary_accuracy: 0.9320
Epoch 3/10
25000/25000 [==============================] - 12s 474us/step - loss: 0.2104 - binary_accuracy: 0.9259 - val_loss: 0.1756 - val_binary_accuracy: 0.9466
Epoch 4/10
25000/25000 [==============================] - 11s 457us/step - loss: 0.1796 - binary_accuracy: 0.9400 - val_loss: 0.1658 - val_binary_accuracy: 0.9422
Epoch 5/10
25000/25000 [==============================] - 11s 450us/step - loss: 0.1594 - binary_accuracy: 0.9456 - val_loss: 0.1340 - val_binary_accuracy: 0.9608
Epoch 6/10
25000/25000 [==============================] - 13s 518us/step - loss: 0.1422 - binary_accuracy: 0.9528 - val_loss: 0.1197 - v

In [39]:
print('Test accuracy: {}'.format(results_1[1]))

Test accuracy: 0.8714


According to the test accuracy, the model with **1 hidden layer** was the best. Continue experimenting with this model.

### Experimenting with the number of hidden units

#### 1 hidden layer of 32 units

In [40]:
model = models.Sequential()
model.add(layers.Dense(32, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss=losses.binary_crossentropy,
              metrics=[metrics.binary_accuracy])

In [41]:
history = model.fit(x_train, y_train, epochs=10, batch_size=512, validation_data=(x_val, y_val))
results = model.evaluate(x_test, y_test)

Train on 25000 samples, validate on 5000 samples
Epoch 1/10
25000/25000 [==============================] - 26s 1ms/step - loss: 0.4239 - binary_accuracy: 0.8245 - val_loss: 0.2623 - val_binary_accuracy: 0.9208
Epoch 2/10
25000/25000 [==============================] - 22s 862us/step - loss: 0.2496 - binary_accuracy: 0.9120 - val_loss: 0.2056 - val_binary_accuracy: 0.9296
Epoch 3/10
25000/25000 [==============================] - 20s 795us/step - loss: 0.1979 - binary_accuracy: 0.9304 - val_loss: 0.1746 - val_binary_accuracy: 0.9402
Epoch 4/10
25000/25000 [==============================] - 16s 655us/step - loss: 0.1691 - binary_accuracy: 0.9410 - val_loss: 0.1382 - val_binary_accuracy: 0.9598
Epoch 5/10
25000/25000 [==============================] - 15s 584us/step - loss: 0.1477 - binary_accuracy: 0.9496 - val_loss: 0.1192 - val_binary_accuracy: 0.9658
Epoch 6/10
25000/25000 [==============================] - 15s 594us/step - loss: 0.1314 - binary_accuracy: 0.9544 - val_loss: 0.1059 - val

In [42]:
print('Test accuracy: {}'.format(results[1]))

Test accuracy: 0.86556


#### 1 hidden layer of 64 units

In [43]:
model = models.Sequential()
model.add(layers.Dense(64, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss=losses.binary_crossentropy,
              metrics=[metrics.binary_accuracy])

In [44]:
history = model.fit(x_train, y_train, epochs=10, batch_size=512, validation_data=(x_val, y_val))
results = model.evaluate(x_test, y_test)

Train on 25000 samples, validate on 5000 samples
Epoch 1/10
25000/25000 [==============================] - 25s 993us/step - loss: 0.4105 - binary_accuracy: 0.8245 - val_loss: 0.2627 - val_binary_accuracy: 0.9084
Epoch 2/10
25000/25000 [==============================] - 19s 756us/step - loss: 0.2384 - binary_accuracy: 0.9134 - val_loss: 0.2385 - val_binary_accuracy: 0.9050
Epoch 3/10
25000/25000 [==============================] - 16s 660us/step - loss: 0.1903 - binary_accuracy: 0.9307 - val_loss: 0.1431 - val_binary_accuracy: 0.9584
Epoch 4/10
25000/25000 [==============================] - 14s 574us/step - loss: 0.1591 - binary_accuracy: 0.9432 - val_loss: 0.1292 - val_binary_accuracy: 0.9596
Epoch 5/10
25000/25000 [==============================] - 15s 586us/step - loss: 0.1423 - binary_accuracy: 0.9488 - val_loss: 0.1243 - val_binary_accuracy: 0.9598
Epoch 6/10
25000/25000 [==============================] - 16s 637us/step - loss: 0.1229 - binary_accuracy: 0.9579 - val_loss: 0.1188 - v

In [45]:
print('Test accuracy: {}'.format(results[1]))

Test accuracy: 0.8608


Having **32 hidden units** seems to work better

### Experimenting with the loss function

**Binary crossentropy** was used before. Accuracy: 0.86556

#### MSE (1 hidden layer of 32 units)

In [48]:
model = models.Sequential()
model.add(layers.Dense(32, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss=losses.mse,
              metrics=[metrics.binary_accuracy])

In [49]:
history = model.fit(x_train, y_train, epochs=10, batch_size=512, validation_data=(x_val, y_val))
results = model.evaluate(x_test, y_test)

Train on 25000 samples, validate on 5000 samples
Epoch 1/10
25000/25000 [==============================] - 22s 864us/step - loss: 0.1363 - binary_accuracy: 0.8280 - val_loss: 0.0798 - val_binary_accuracy: 0.9210
Epoch 2/10
25000/25000 [==============================] - 17s 686us/step - loss: 0.0755 - binary_accuracy: 0.9144 - val_loss: 0.0588 - val_binary_accuracy: 0.9374
Epoch 3/10
25000/25000 [==============================] - 15s 597us/step - loss: 0.0592 - binary_accuracy: 0.9321 - val_loss: 0.0479 - val_binary_accuracy: 0.9496
Epoch 4/10
25000/25000 [==============================] - 13s 515us/step - loss: 0.0501 - binary_accuracy: 0.9421 - val_loss: 0.0384 - val_binary_accuracy: 0.9642
Epoch 5/10
25000/25000 [==============================] - 14s 548us/step - loss: 0.0429 - binary_accuracy: 0.9514 - val_loss: 0.0402 - val_binary_accuracy: 0.9566
Epoch 6/10
25000/25000 [==============================] - 14s 542us/step - loss: 0.0376 - binary_accuracy: 0.9596 - val_loss: 0.0300 - v

In [50]:
print('Test accuracy: {}'.format(results[1]))

Test accuracy: 0.86652


The model using the **mse** loss function performs a little better

### Experimenting with activation

**Relu** was used before. Accuracy: 0.86652

#### Tanh activation

In [51]:
model = models.Sequential()
model.add(layers.Dense(32, activation='tanh', input_shape=(10000,)))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss=losses.mse,
              metrics=[metrics.binary_accuracy])

In [52]:
history = model.fit(x_train, y_train, epochs=10, batch_size=512, validation_data=(x_val, y_val))
results = model.evaluate(x_test, y_test)

Train on 25000 samples, validate on 5000 samples
Epoch 1/10
25000/25000 [==============================] - 21s 838us/step - loss: 0.1335 - binary_accuracy: 0.8290 - val_loss: 0.0760 - val_binary_accuracy: 0.9232
Epoch 2/10
25000/25000 [==============================] - 16s 642us/step - loss: 0.0735 - binary_accuracy: 0.9121 - val_loss: 0.0525 - val_binary_accuracy: 0.9434
Epoch 3/10
25000/25000 [==============================] - 16s 638us/step - loss: 0.0552 - binary_accuracy: 0.9339 - val_loss: 0.0519 - val_binary_accuracy: 0.9362
Epoch 4/10
25000/25000 [==============================] - 14s 553us/step - loss: 0.0454 - binary_accuracy: 0.9469 - val_loss: 0.0322 - val_binary_accuracy: 0.9694
Epoch 5/10
25000/25000 [==============================] - 15s 588us/step - loss: 0.0386 - binary_accuracy: 0.9550 - val_loss: 0.0351 - val_binary_accuracy: 0.9594
Epoch 6/10
25000/25000 [==============================] - 13s 527us/step - loss: 0.0333 - binary_accuracy: 0.9623 - val_loss: 0.0317 - v

In [53]:
print('Test accuracy: {}'.format(results[1]))

Test accuracy: 0.85852


The model that uses the **relu** activation works a little better

# Part 2: classifying newswires

Reuters dataset

In [63]:
from keras.datasets import reuters

(train_data, train_labels), (test_data, test_labels) = reuters.load_data(
    num_words=10000)

In [64]:
def vectorize_sequences(sequences, dimension=10000):
    results = np.zeros((len(sequences), dimension))
    for i, sequence in enumerate(sequences):
        results[i, sequence] = 1.
    return results

x_train = vectorize_sequences(train_data)
x_test = vectorize_sequences(test_data)

In [65]:
from keras.utils.np_utils import to_categorical

one_hot_train_labels = to_categorical(train_labels)
one_hot_test_labels = to_categorical(test_labels)

In [66]:
model = models.Sequential()
model.add(layers.Dense(64, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(46, activation='softmax'))

model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss=losses.categorical_crossentropy,
              metrics=[metrics.binary_accuracy])

In [67]:
x_val = x_train[:1000]
partial_x_train = x_train[1000:]

y_val = one_hot_train_labels[:1000]
partial_y_train = one_hot_train_labels[1000:]

In [68]:
history = model.fit(partial_x_train,
                    partial_y_train,
                    epochs=9,
                    batch_size=512,
                    validation_data=(x_val, y_val)
                   )

Train on 7982 samples, validate on 1000 samples
Epoch 1/9
7982/7982 [==============================] - 5s 566us/step - loss: 2.5353 - binary_accuracy: 0.9805 - val_loss: 1.7273 - val_binary_accuracy: 0.9852
Epoch 2/9
7982/7982 [==============================] - 3s 403us/step - loss: 1.4522 - binary_accuracy: 0.9878 - val_loss: 1.3517 - val_binary_accuracy: 0.9890
Epoch 3/9
7982/7982 [==============================] - 2s 311us/step - loss: 1.1012 - binary_accuracy: 0.9906 - val_loss: 1.1743 - val_binary_accuracy: 0.9902
Epoch 4/9
7982/7982 [==============================] - 2s 300us/step - loss: 0.8745 - binary_accuracy: 0.9922 - val_loss: 1.0851 - val_binary_accuracy: 0.9907
Epoch 5/9
7982/7982 [==============================] - 2s 305us/step - loss: 0.7074 - binary_accuracy: 0.9933 - val_loss: 0.9864 - val_binary_accuracy: 0.9908
Epoch 6/9
7982/7982 [==============================] - 2s 310us/step - loss: 0.5697 - binary_accuracy: 0.9944 - val_loss: 0.9403 - val_binary_accuracy: 0.991

In [70]:
results = model.evaluate(x_test, one_hot_test_labels)

2246/2246 [==============================] - 1s 316us/step


In [71]:
results[1]

0.9911920869339819

#### Using less hidden units (32)

In [72]:
model = models.Sequential()
model.add(layers.Dense(32, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(32, activation='relu'))
model.add(layers.Dense(46, activation='softmax'))

model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss=losses.categorical_crossentropy,
              metrics=[metrics.binary_accuracy])

history = model.fit(partial_x_train,
                    partial_y_train,
                    epochs=9,
                    batch_size=512,
                    validation_data=(x_val, y_val)
                   )
results = model.evaluate(x_test, one_hot_test_labels)

Train on 7982 samples, validate on 1000 samples
Epoch 1/9
7982/7982 [==============================] - 7s 835us/step - loss: 3.0975 - binary_accuracy: 0.9785 - val_loss: 2.3174 - val_binary_accuracy: 0.9797
Epoch 2/9
7982/7982 [==============================] - 3s 343us/step - loss: 1.8980 - binary_accuracy: 0.9827 - val_loss: 1.6703 - val_binary_accuracy: 0.9854
Epoch 3/9
7982/7982 [==============================] - 2s 311us/step - loss: 1.4047 - binary_accuracy: 0.9878 - val_loss: 1.4038 - val_binary_accuracy: 0.9884
Epoch 4/9
7982/7982 [==============================] - 2s 274us/step - loss: 1.1582 - binary_accuracy: 0.9905 - val_loss: 1.2635 - val_binary_accuracy: 0.9899
Epoch 5/9
7982/7982 [==============================] - 3s 324us/step - loss: 0.9933 - binary_accuracy: 0.9916 - val_loss: 1.1801 - val_binary_accuracy: 0.9904
Epoch 6/9
7982/7982 [==============================] - 3s 350us/step - loss: 0.8638 - binary_accuracy: 0.9925 - val_loss: 1.1192 - val_binary_accuracy: 0.990

In [73]:
results[1]

0.9907274944168696

#### Using less hidden units (16)

In [76]:
model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dense(46, activation='softmax'))

model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss=losses.categorical_crossentropy,
              metrics=[metrics.binary_accuracy])

history = model.fit(partial_x_train,
                    partial_y_train,
                    epochs=9,
                    batch_size=512,
                    validation_data=(x_val, y_val)
                   )
results = model.evaluate(x_test, one_hot_test_labels)

Train on 7982 samples, validate on 1000 samples
Epoch 1/9
7982/7982 [==============================] - 3s 394us/step - loss: 3.4649 - binary_accuracy: 0.9783 - val_loss: 3.0572 - val_binary_accuracy: 0.9784
Epoch 2/9
7982/7982 [==============================] - 2s 282us/step - loss: 2.7043 - binary_accuracy: 0.9789 - val_loss: 2.3974 - val_binary_accuracy: 0.9796
Epoch 3/9
7982/7982 [==============================] - 2s 272us/step - loss: 2.0849 - binary_accuracy: 0.9813 - val_loss: 1.9039 - val_binary_accuracy: 0.9828
Epoch 4/9
7982/7982 [==============================] - 2s 297us/step - loss: 1.6596 - binary_accuracy: 0.9850 - val_loss: 1.6010 - val_binary_accuracy: 0.9860
Epoch 5/9
7982/7982 [==============================] - 2s 255us/step - loss: 1.3896 - binary_accuracy: 0.9880 - val_loss: 1.4155 - val_binary_accuracy: 0.9877
Epoch 6/9
7982/7982 [==============================] - 2s 259us/step - loss: 1.2076 - binary_accuracy: 0.9897 - val_loss: 1.3008 - val_binary_accuracy: 0.989

In [77]:
results[1]

0.9902822586541078

#### Using more hidden units (128)

In [74]:
model = models.Sequential()
model.add(layers.Dense(128, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(128, activation='relu'))
model.add(layers.Dense(46, activation='softmax'))

model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss=losses.categorical_crossentropy,
              metrics=[metrics.binary_accuracy])

history = model.fit(partial_x_train,
                    partial_y_train,
                    epochs=9,
                    batch_size=512,
                    validation_data=(x_val, y_val)
                   )
results = model.evaluate(x_test, one_hot_test_labels)

Train on 7982 samples, validate on 1000 samples
Epoch 1/9
7982/7982 [==============================] - 4s 525us/step - loss: 2.2531 - binary_accuracy: 0.9822 - val_loss: 1.4524 - val_binary_accuracy: 0.9874
Epoch 2/9
7982/7982 [==============================] - 4s 456us/step - loss: 1.1717 - binary_accuracy: 0.9901 - val_loss: 1.1415 - val_binary_accuracy: 0.9901
Epoch 3/9
7982/7982 [==============================] - 3s 386us/step - loss: 0.8202 - binary_accuracy: 0.9925 - val_loss: 1.0116 - val_binary_accuracy: 0.9907
Epoch 4/9
7982/7982 [==============================] - 3s 365us/step - loss: 0.5955 - binary_accuracy: 0.9941 - val_loss: 0.9961 - val_binary_accuracy: 0.9903
Epoch 5/9
7982/7982 [==============================] - 4s 440us/step - loss: 0.4333 - binary_accuracy: 0.9954 - val_loss: 0.9036 - val_binary_accuracy: 0.9915
Epoch 6/9
7982/7982 [==============================] - 3s 423us/step - loss: 0.3378 - binary_accuracy: 0.9963 - val_loss: 0.8720 - val_binary_accuracy: 0.992

In [75]:
results[1]

0.9903596871576453

To avoid overloading the model, let's continue using 32 units

#### Using 1 hidden layer

In [78]:
model = models.Sequential()
model.add(layers.Dense(32, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(46, activation='softmax'))

model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss=losses.categorical_crossentropy,
              metrics=[metrics.binary_accuracy])

history = model.fit(partial_x_train,
                    partial_y_train,
                    epochs=9,
                    batch_size=512,
                    validation_data=(x_val, y_val)
                   )
results = model.evaluate(x_test, one_hot_test_labels)

Train on 7982 samples, validate on 1000 samples
Epoch 1/9
7982/7982 [==============================] - 3s 397us/step - loss: 3.1273 - binary_accuracy: 0.9784 - val_loss: 2.4809 - val_binary_accuracy: 0.9789
Epoch 2/9
7982/7982 [==============================] - 2s 291us/step - loss: 2.0967 - binary_accuracy: 0.9805 - val_loss: 1.8341 - val_binary_accuracy: 0.9826
Epoch 3/9
7982/7982 [==============================] - 2s 282us/step - loss: 1.5477 - binary_accuracy: 0.9849 - val_loss: 1.4802 - val_binary_accuracy: 0.9860
Epoch 4/9
7982/7982 [==============================] - 2s 283us/step - loss: 1.2207 - binary_accuracy: 0.9883 - val_loss: 1.2708 - val_binary_accuracy: 0.9879
Epoch 5/9
7982/7982 [==============================] - 2s 286us/step - loss: 1.0067 - binary_accuracy: 0.9906 - val_loss: 1.1436 - val_binary_accuracy: 0.9898
Epoch 6/9
7982/7982 [==============================] - 2s 291us/step - loss: 0.8518 - binary_accuracy: 0.9922 - val_loss: 1.0562 - val_binary_accuracy: 0.990

In [79]:
results[1]

0.9912695226558913

#### Using 3 hidden layers

In [80]:
model = models.Sequential()
model.add(layers.Dense(32, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(32, activation='relu'))
model.add(layers.Dense(32, activation='relu'))
model.add(layers.Dense(46, activation='softmax'))

model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss=losses.categorical_crossentropy,
              metrics=[metrics.binary_accuracy])

history = model.fit(partial_x_train,
                    partial_y_train,
                    epochs=9,
                    batch_size=512,
                    validation_data=(x_val, y_val)
                   )
results = model.evaluate(x_test, one_hot_test_labels)

Train on 7982 samples, validate on 1000 samples
Epoch 1/9
7982/7982 [==============================] - 4s 465us/step - loss: 3.1505 - binary_accuracy: 0.9783 - val_loss: 2.3734 - val_binary_accuracy: 0.9791
Epoch 2/9
7982/7982 [==============================] - 2s 309us/step - loss: 1.9585 - binary_accuracy: 0.9824 - val_loss: 1.6551 - val_binary_accuracy: 0.9859
Epoch 3/9
7982/7982 [==============================] - 3s 326us/step - loss: 1.4305 - binary_accuracy: 0.9883 - val_loss: 1.3745 - val_binary_accuracy: 0.9890
Epoch 4/9
7982/7982 [==============================] - 2s 289us/step - loss: 1.1689 - binary_accuracy: 0.9904 - val_loss: 1.2751 - val_binary_accuracy: 0.9897
Epoch 5/9
7982/7982 [==============================] - 2s 286us/step - loss: 0.9978 - binary_accuracy: 0.9915 - val_loss: 1.1999 - val_binary_accuracy: 0.9900
Epoch 6/9
7982/7982 [==============================] - 2s 284us/step - loss: 0.8617 - binary_accuracy: 0.9924 - val_loss: 1.1271 - val_binary_accuracy: 0.990

In [81]:
results[1]

0.9906307068767021